src/type_conversion/jsimport.c

```diff
@@ -1,77 +1,350 @@
 #include "jsimport.h"
+#include "Python.h"
+#include "jsproxy.h"
+#include "stdbool.h"
 
 #include <emscripten.h>
 
 #include "hiwire.h"
 #include "js2python.h"
 
-static PyObject* js_module = NULL;
+_Py_IDENTIFIER(__dict__);
+_Py_IDENTIFIER(__dir__);
+_Py_IDENTIFIER(get);
+_Py_IDENTIFIER(keys);
+
+static void
+_JsImport_setHiwireObject(PyObject* module, int id)
+{
+  PyObject** proxy = (PyObject**)PyModule_GetState(module);
+  // JsProxy_cnew can memory error but that's the only error it throws.
+  // Our current plan is to crash and burn if that ever happens.
+  *proxy = JsProxy_cnew(id);
+}
 
 static PyObject*
-JsImport_GetAttr(PyObject* self, PyObject* attr)
+_JsImport_getJsProxy(PyObject* module)
 {
-  const char* c = PyUnicode_AsUTF8(attr);
-  if (c == NULL) {
-    return NULL;
+  PyObject** proxy = (PyObject**)PyModule_GetState(module);
+  return *proxy;
+}
+
+// Annoyingly the __dir__ method does NOT get a self parameter.
+// We need to wrap it in a closure with a reference to our js object.
+// First we make the Python type that does this closure.
+
+// clang-format off
+typedef struct
+{
+  PyObject_HEAD
+  PyObject* module;
+} JsImportDir;
+// clang-format on
+
+static bool
+is_jsproxy_map(PyObject* proxy)
+{
+  PyObject* getfunc = _PyObject_GetAttrId(proxy, &PyId_get);
+  if (getfunc) {
+    Py_CLEAR(getfunc);
+    return true;
+  } else {
+    PyErr_Clear();
+    return false;
+  }
+}
+
+static int
+JsImportDir_init(PyObject* o, PyObject* args, PyObject* kwargs)
+{
+  JsImportDir* self = (JsImportDir*)o;
+  PyObject* module;
+  if (!PyArg_UnpackTuple(args, "__init__", 1, 1, &module)) {
+    return -1;
   }
-  int idval = hiwire_get_global((int)c);
```

Comment:
These should be the only uses of hiwire_get_global. You can delete that function together with this PR.

Reply:
Well I reverted some changes. Removing this and `JsBoundMethod` and fixing the bugs will have to wait for another PR.